src/algorithm/HISTOGRAM-Cuda.cpp

```diff
@@ -0,0 +1,288 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-24, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "HISTOGRAM.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#if defined(RAJA_ENABLE_CUDA)
+
+#include "cub/device/device_histogram.cuh"
+#include "cub/util_allocator.cuh"
+
+#include "common/CudaDataUtils.hpp"
+
+#include <iostream>
+
+namespace rajaperf
+{
+namespace algorithm
+{
+
+template < size_t block_size, size_t replication >
+__launch_bounds__(block_size)
+__global__ void histogram(HISTOGRAM::Data_ptr counts,
+                          Index_ptr bins,
+                          Index_type iend)
+{
+  Index_type i = blockIdx.x * block_size + threadIdx.x;
+  if (i < iend) {
+    HISTOGRAM_GPU_RAJA_BODY(RAJA::cuda_atomic);
+  }
+}
+
+
+void HISTOGRAM::runCudaVariantLibrary(VariantID vid)
+{
+  const Index_type run_reps = getRunReps();
+  const Index_type ibegin = 0;
+  const Index_type iend = getActualProblemSize();
+
+  auto res{getCudaResource()};
+
+  HISTOGRAM_GPU_DATA_SETUP;
+
+  RAJAPERF_CUDA_REDUCER_SETUP(Data_ptr, counts, hcounts, num_bins, 1);
+
+  RAJAPERF_UNUSED_VAR(counts_init);
+
+  if ( vid == Base_CUDA ) {
+
+    cudaStream_t stream = res.get_stream();
+
+    int len = iend - ibegin;
+
+    // Determine temporary device storage requirements
+    void* d_temp_storage = nullptr;
+    size_t temp_storage_bytes = 0;
+    cudaErrchk(::cub::DeviceHistogram::HistogramEven(d_temp_storage,
+                                                     temp_storage_bytes,
+                                                     bins+ibegin,
+                                                     counts,
+                                                     static_cast<int>(num_bins+1),
+                                                     static_cast<Index_type>(0),
+                                                     num_bins,
+                                                     len,
+                                                     stream));
+
+    // Allocate temporary storage
+    unsigned char* temp_storage;
+    allocData(DataSpace::CudaDevice, temp_storage, temp_storage_bytes);
+    d_temp_storage = temp_storage;
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      // Run
```

Comment:
go kernel go!

Reply:
Let's remove this pointless comment.

Reply:
Never-mind, this is with cub/rocprim stuff that is a bit confusing about when it actually runs.